### PR TITLE
Add WFS Client Test

### PR DESCRIPTION
**Pull request by geographika**
While setting up a layer using the [WFS Client](https://www.mapserver.org/ogc/wfs_client.html) I noticed there were no tests in place. 

I've added one (that cascades a WFS service) that works fine locally but fails in the CI. Failing logs as follows:

```
+[Thu Jun 15 13:04:38 2023].66359 HTTP: Before download loop
+[Thu Jun 15 13:04:38 2023].433103 msHTTPWriteFct(id=0, 2496 bytes)
+[Thu Jun 15 13:04:38 2023].433165 HTTP: After download loop
+[Thu Jun 15 13:04:38 2023].433176 msHTTPExecuteRequests() timing summary per layer (connect_time + time_to_first_packet + download_time = total_time in seconds)
+[Thu Jun 15 13:04:38 2023].433242 Layer 0: 0.082 + 0.285 + 0.082 = 0.449s
+[Thu Jun 15 13:04:38 2023].435265 msOGRFileOpen(/tmp/648b0c66_a1b5_5.tmp.gml)...
+[Thu Jun 15 13:04:38 2023].435303 OGROPen(/tmp/648b0c66_a1b5_5.tmp.gml)
+[Thu Jun 15 13:04:38 2023].816213 msConnPoolRegister(continents,https://demo.mapserver.org/cgi-bin/wfs?,0x5642047d0220)
+[Thu Jun 15 13:04:38 2023].816273 msOGRFileWhichShapes: Setting spatial filter to -180 -90 180 90
+[Thu Jun 15 13:04:38 2023].816337 msOGRFileWhichShapes: Setting spatial filter to -86.92396 46.381586 -60.381746 65.176486
+[Thu Jun 15 13:04:38 2023].816521 msOGRFileNextShape: Returning shape=23774, tile=0
+[Thu Jun 15 13:04:38 2023].816556 msOGRFileNextShape: Returning shape=18417, tile=0
+[Thu Jun 15 13:04:38 2023].816575 msOGRFileNextShape(): OGR error. OGR GetNextFeature() error'd. Check logs.
+[Thu Jun 15 13:04:38 2023].816582 msOGRFileNextShape(): HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1)
+[Thu Jun 15 13:04:38 2023].816591 msWFSGetFeature(): WFS server error. ms_error->code not found
```

I'm thinking it could be a SSL/Cert issue, but the WMS Client uses the same server and the tests work fine.

### CODE REVIEW

**Comment by geographika** [comment]
I think the errors are very likely related to the HTTP/2 issues with the demo server mentioned in https://github.com/MapServer/MapServer/pull/6857

There are also memory leaks when running the code - see https://app.travis-ci.com/github/MapServer/MapServer/jobs/604378829#L5236

**Comment by geographika** [comment]
For reference when the Travis logs are removed, memory leak trace:

```
=================================================================
==39300==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 48 byte(s) in 3 object(s) allocated from:
    #0 0x7fce0175d808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x7fcdfd1907cd  (/lib/x86_64-linux-gnu/libcurl-gnutls.so.4+0x417cd)
Indirect leak of 23 byte(s) in 1 object(s) allocated from:
    #0 0x7fce016e63ed in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cc:445
    #1 0x7fcdfd190829 in curl_slist_append (/lib/x86_64-linux-gnu/libcurl-gnutls.so.4+0x41829)
    #2 0x7fce002ccc6c in msOWSExecuteRequests /home/travis/build/MapServer/MapServer/mapows.c:2521
    #3 0x7fce00948f37 in msWFSLayerWhichShapes /home/travis/build/MapServer/MapServer/mapwfslayer.c:1101
    #4 0x7fce00947d29 in msWFSLayerOpen /home/travis/build/MapServer/MapServer/mapwfslayer.c:811
    #5 0x7fce00947d9e in msWFSLayerOpenVT /home/travis/build/MapServer/MapServer/mapwfslayer.c:835
    #6 0x7fce006823d9 in msLayerOpen /home/travis/build/MapServer/MapServer/maplayer.c:248
    #7 0x7fce00683ec9 in msLayerGetExtent /home/travis/build/MapServer/MapServer/maplayer.c:556
    #8 0x7fce002ccbc1 in msOWSGetLayerExtent /home/travis/build/MapServer/MapServer/mapows.c:2501
    #9 0x7fce008daacb in msWFSRunBasicGetFeature /home/travis/build/MapServer/MapServer/mapwfs.cpp:2208
    #10 0x7fce008de2fe in msWFSRetrieveFeatures /home/travis/build/MapServer/MapServer/mapwfs.cpp:2551
    #11 0x7fce008ebaad in msWFSGetFeature /home/travis/build/MapServer/MapServer/mapwfs.cpp:3830
    #12 0x7fce008f2f1c in msWFSDispatch /home/travis/build/MapServer/MapServer/mapwfs.cpp:4637
    #13 0x7fce002bfd59 in msOWSDispatch /home/travis/build/MapServer/MapServer/mapows.c:307
    #14 0x7fce00535236 in msCGIDispatchRequest /home/travis/build/MapServer/MapServer/mapservutil.c:1919
    #15 0x557defd67f00 in main /home/travis/build/MapServer/MapServer/mapserv.c:288
    #16 0x7fcdfca9e082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
Indirect leak of 23 byte(s) in 1 object(s) allocated from:
    #0 0x7fce016e63ed in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cc:445
    #1 0x7fcdfd190829 in curl_slist_append (/lib/x86_64-linux-gnu/libcurl-gnutls.so.4+0x41829)
    #2 0x7fce002ccc6c in msOWSExecuteRequests /home/travis/build/MapServer/MapServer/mapows.c:2521
    #3 0x7fce00948f37 in msWFSLayerWhichShapes /home/travis/build/MapServer/MapServer/mapwfslayer.c:1101
    #4 0x7fce00947d29 in msWFSLayerOpen /home/travis/build/MapServer/MapServer/mapwfslayer.c:811
    #5 0x7fce00947d9e in msWFSLayerOpenVT /home/travis/build/MapServer/MapServer/mapwfslayer.c:835
    #6 0x7fce006823d9 in msLayerOpen /home/travis/build/MapServer/MapServer/maplayer.c:248
    #7 0x7fce00736dab in msQueryByRect /home/travis/build/MapServer/MapServer/mapquery.c:1122
    #8 0x7fce008db433 in msWFSRunBasicGetFeature /home/travis/build/MapServer/MapServer/mapwfs.cpp:2260
    #9 0x7fce008de2fe in msWFSRetrieveFeatures /home/travis/build/MapServer/MapServer/mapwfs.cpp:2551
    #10 0x7fce008ebaad in msWFSGetFeature /home/travis/build/MapServer/MapServer/mapwfs.cpp:3830
    #11 0x7fce008f2f1c in msWFSDispatch /home/travis/build/MapServer/MapServer/mapwfs.cpp:4637
    #12 0x7fce002bfd59 in msOWSDispatch /home/travis/build/MapServer/MapServer/mapows.c:307
    #13 0x7fce00535236 in msCGIDispatchRequest /home/travis/build/MapServer/MapServer/mapservutil.c:1919
    #14 0x557defd67f00 in main /home/travis/build/MapServer/MapServer/mapserv.c:288
    #15 0x7fcdfca9e082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
Indirect leak of 23 byte(s) in 1 object(s) allocated from:
    #0 0x7fce016e63ed in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cc:445
    #1 0x7fcdfd190829 in curl_slist_append (/lib/x86_64-linux-gnu/libcurl-gnutls.so.4+0x41829)
    #2 0x7fce002ccc6c in msOWSExecuteRequests /home/travis/build/MapServer/MapServer/mapows.c:2521
    #3 0x7fce00948f37 in msWFSLayerWhichShapes /home/travis/build/MapServer/MapServer/mapwfslayer.c:1101
    #4 0x7fce00947d29 in msWFSLayerOpen /home/travis/build/MapServer/MapServer/mapwfslayer.c:811
    #5 0x7fce00947d9e in msWFSLayerOpenVT /home/travis/build/MapServer/MapServer/mapwfslayer.c:835
    #6 0x7fce006823d9 in msLayerOpen /home/travis/build/MapServer/MapServer/maplayer.c:248
    #7 0x7fce008e68b6 in msWFSGetFeature /home/travis/build/MapServer/MapServer/mapwfs.cpp:3504
    #8 0x7fce008f2f1c in msWFSDispatch /home/travis/build/MapServer/MapServer/mapwfs.cpp:4637
    #9 0x7fce002bfd59 in msOWSDispatch /home/travis/build/MapServer/MapServer/mapows.c:307
    #10 0x7fce00535236 in msCGIDispatchRequest /home/travis/build/MapServer/MapServer/mapservutil.c:1919
    #11 0x557defd67f00 in main /home/travis/build/MapServer/MapServer/mapserv.c:288
    #12 0x7fcdfca9e082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
SUMMARY: AddressSanitizer: 117 byte(s) leaked in 6 allocation(s).
make: *** [Makefile:93: asan_compatible_tests] Error 1
```

**Comment by geographika** [comment]
With the fixes in #6954 to the demo server the request is made, however the test results in the CI aren't what I'm getting locally. 
The header `Content-Type: text/xml; charset=UTF-8` seems to be stripped in the CI, even though the test isn't meant to remove it. 

The memory leak is gone so it appears as this only occurs when the HTTP request fails.

**Comment by geographika** [comment]
The memory leak is now appears to be fixed by adding the `curl_slist_free_all(headers)` call at the end of the function after requests have been made. I'd appreciate a review though before merging as I don't fully understand the implications of using `curl_multi_perform` with the same header list.

**Comment by rouault** [comment]
> I'd appreciate a review though before merging as I don't fully understand the implications of using `curl_multi_perform` with the same header list.

I don't think your proposed fix is safe. The safest thing to do would be to add a ``headers`` member to pasReqInfo[], and have thus a per-request header list. And on cleanup you would iterate over it to call curl_slist_free_all(pasReqInfo[i].headers)

File: src/maphttp.c [hunk 1]
```diff
@@ -484,6 +484,9 @@ int msHTTPExecuteRequests(httpRequestObj *pasReqInfo, int numRequests,
     return (MS_FAILURE);
   }
 
+  struct curl_slist **headersArray =
+      msSmallMalloc(numRequests * sizeof(struct curl_slist *));
```
Comment (rouault) [inline src/maphttp.c:488]: to make sure entries are null initialized:
```suggestion
      msSmallCalloc(numRequests, sizeof(struct curl_slist *));
```

File: src/maphttp.c [hunk 2]
```diff
@@ -914,6 +915,13 @@ int msHTTPExecuteRequests(httpRequestObj *pasReqInfo, int numRequests,
     /* Cleanup this handle */
     unchecked_curl_easy_setopt(http_handle, CURLOPT_URL, "");
     curl_multi_remove_handle(multi_handle, http_handle);
+
+    for (int i = 0; i < numRequests; ++i) {
+      curl_slist_free_all(headersArray[i]);
+    }
+
+    msFree(headersArray);
+
```
Comment (rouault) [inline src/maphttp.c:924]: that whole block should be moved after line 927.

**Comment by geographika** [comment]
Thanks for your suggestions and feedback @rouault.

I switched to your original proposal of adding a new `curl_headers` property to `httpRequestObj` to store the headers for each request. I thought as `curl_slist` was a cURL type and cURL is an optional dependency I'd have to use a workaround, but then noticed the `curl_handle` property has a type of `void` so took the same approach. 

There are several cleanup functions such as `curl_easy_cleanup` at the end of `msHTTPExecuteRequests`. There are also several `return` statements for various errors prior to that. I presume this means no cleanup is done if the requests fail. These are handled using `goto` in other parts of the MapServer codebase - is this the recommended approach (any changes relating to this would be outside the scope of this pull request).

**Comment by rouault** [comment]
> These are handled using `goto` in other parts of the MapServer codebase - is this the recommended approach (any changes relating to this would be outside the scope of this pull request).

``goto`` shouldn't be abused, but for a C-style code it is somewhat aceptatble, as there isn't unfortunately much sexier alternatives, other than sometimes putting the code in a ``do {  .... } while(0);`` dummy loop, and using ``break`` to exit it early in case of errors. But that's a bit a ``goto`` strategy in disguise... In a pure C++ approach with objects with destructors, cleanups are much easier to do. Agreed this is out of scope of this PR

**Pull request merged**